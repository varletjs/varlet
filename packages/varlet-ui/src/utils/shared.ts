import { TimeData } from '../countdown/props'

export interface CacheInstance<T> {
  cache: T[]

  has(key: T): boolean

  add(key: T): void

  remove(key: T): void

  clear(): void
}

export const toNumber = (val: number | string | boolean | undefined | null): number => {
  if (val == null) return 0

  if (isString(val)) {
    val = parseFloat(val)
    val = Number.isNaN(val) ? 0 : val
    return val
  }

  if (isBool(val)) return Number(val)

  return val
}

export const isHTMLSupportImage = (val: string | undefined | null) => {
  if (val == null) {
    return false
  }

  return val.startsWith('data:image') || /\.(png|jpg|gif|jpeg|svg)$/.test(val)
}

export const isHTMLSupportVideo = (val: string | undefined | null) => {
  if (val == null) {
    return false
  }

  return val.startsWith('data:video') || /\.(mp4|webm|ogg)$/.test(val)
}

export const isString = (val: unknown): val is string => typeof val === 'string'

export const isBool = (val: unknown): val is boolean => typeof val === 'boolean'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isPlainObject = (val: unknown): val is Record<string, any> =>
  Object.prototype.toString.call(val) === '[object Object]'

export const isObject = (val: unknown): val is Record<string, any> => typeof val === 'object' && val !== null

export const isArray = (val: unknown): val is Array<any> => Array.isArray(val)

export const isURL = (val: string | undefined | null) => {
  if (!val) {
    return false
  }

  return /^(http)|(\.*\/)/.test(val)
}

export const isEmpty = (val: unknown) =>
  val === undefined || val === null || val === '' || (Array.isArray(val) && !val.length)

export const removeItem = (arr: Array<unknown>, item: unknown) => {
  if (arr.length) {
    const index: number = arr.indexOf(item)
    if (index > -1) {
      return arr.splice(index, 1)
    }
  }
}

export const throttle = (method: any, mustRunDelay = 200): (() => void) => {
  let timer: number
  let start = 0

  return function loop(this: unknown, ...args) {
    const now = Date.now()
    const elapsed = now - start

    if (!start) {
      start = now
    }

    if (timer) {
      window.clearTimeout(timer)
    }

    if (elapsed >= mustRunDelay) {
      method.apply(this, args)
      start = now
    } else {
      timer = window.setTimeout(() => {
        loop.apply(this, args)
      }, mustRunDelay - elapsed)
    }
  }
}

export const createCache = <T>(max: number): CacheInstance<T> => {
  const cache: T[] = []

  return {
    cache,

    has(key: T) {
      return this.cache.includes(key)
    },

    add(key: T) {
      if (this.has(key)) {
        return
      }

      this.cache.length === max && cache.shift()
      this.cache.push(key)
    },

    remove(key: T) {
      this.has(key) && removeItem(this.cache, key)
    },

    clear() {
      this.cache.length = 0
    },
  }
}

export const linear = (value: number): number => value

export const cubic = (value: number): number => value ** 3

export const easeInOutCubic = (value: number): number =>
  value < 0.5 ? cubic(value * 2) / 2 : 1 - cubic((1 - value) * 2) / 2

export function parseFormat(format: string, time: TimeData): string {
  const scannedTimes = Object.values(time)
  const scannedFormats = ['DD', 'HH', 'mm', 'ss']
  const padValues = [24, 60, 60, 1000]

  scannedFormats.forEach((scannedFormat, index) => {
    if (!format.includes(scannedFormat)) {
      scannedTimes[index + 1] += scannedTimes[index] * padValues[index]
    } else {
      format = format.replace(scannedFormat, String(scannedTimes[index]).padStart(2, '0'))
    }
  })

  if (format.includes('S')) {
    const ms = String(scannedTimes[scannedTimes.length - 1]).padStart(3, '0')

    if (format.includes('SSS')) {
      format = format.replace('SSS', ms)
    } else if (format.includes('SS')) {
      format = format.replace('SS', ms.slice(0, 2))
    } else {
      format = format.replace('S', ms.slice(0, 1))
    }
  }

  return format
}

export const dt = (value: unknown, defaultText: string | undefined) => (value == null ? defaultText : value)

export const inBrowser = () => typeof window !== 'undefined'

export const uniq = (arr: Array<any>) => [...new Set(arr)]

export function kebabCase(str: string): string {
  const ret = str.replace(/([A-Z])/g, ' $1').trim()
  return ret.split(' ').join('-').toLowerCase()
}
