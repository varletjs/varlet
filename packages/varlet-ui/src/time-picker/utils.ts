import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
