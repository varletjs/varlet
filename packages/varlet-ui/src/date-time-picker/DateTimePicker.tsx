import './dateTimePicker.less'
import '../button/button.less'
import '../tab-item/tabItem.less'

import { defineComponent, VNodeChild, watch } from 'vue'
import { props } from './props'
import { call } from '@varlet/shared'
import { useVModel } from '@varlet/use'
import { createNamespace, flatFragment } from '../utils/components'
import VarButton from '../button'
import VarTabsItems from '../tabs-items'
import VarTabItem from '../tab-item'

const { name, n, classes } = createNamespace('date-time-picker')

export default defineComponent({
  name,
  props,
  setup(props, { slots }) {
    const activeTab = useVModel(props, 'active')
    watch(activeTab, () => {})
    const cancel = () => {
      call(props.onCancel)
    }
    let children: VNodeChild[] = call(slots.default) ?? []
    children = flatFragment(children)
    const nextStep = () => {
      activeTab.value += 1
    }
    const prevStep = () => {
      if (activeTab.value) {
        activeTab.value -= 1
      }
    }
    const timeConfirm = () => {
      call(props.onConfirm)
    }

    return () => (
      <div class={classes(n())} ref="picker">
        <VarTabsItems v-model:active={activeTab.value}>
          {children.map((child) => (
            <VarTabItem>{child}</VarTabItem>
          ))}
        </VarTabsItems>
        <div class={n('body')}>
          <div class={n('button-flex')}>
            <VarButton class={n('prev-button')} text onClick={prevStep} v-show={activeTab.value}>
              {props.prevStepText}
            </VarButton>
            <VarButton class={n('next-button')} text onClick={nextStep} v-show={activeTab.value < children.length - 1}>
              {props.nextStepText}
            </VarButton>
          </div>
          <div class={n('button-flex')}>
            <VarButton class={n('cancel-button')} text onClick={cancel}>
              {props.cancelButtonText}
            </VarButton>
            <VarButton class={n('confirm-button')} text type="primary" onClick={timeConfirm}>
              {props.confirmButtonText}
            </VarButton>
          </div>
        </div>
      </div>
    )
  },
})
