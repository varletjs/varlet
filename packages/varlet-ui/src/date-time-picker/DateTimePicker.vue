<template>
  <div :class="classes(n(), formatElevation(elevation, 2))" ref="picker">
    <var-tabs-items v-model:active="activeTab">
      <var-tab-item>
        <var-date-picker v-model="date1"> </var-date-picker>
        <div :class="n('body')">
          <var-button text @click="cancel"> {{ cancelButtonText }}</var-button>
          <var-button text type="primary" @click="nextStep"> {{ nextStepText }}</var-button>
        </div>
      </var-tab-item>
      <var-tab-item>
        <var-time-picker v-model="date2"> </var-time-picker>
        <div :class="n('body')">
          <var-button text type="primary" @click="prevStep"> {{ prevStepText }} </var-button>
          <var-button text @click="cancel"> {{ cancelButtonText }}</var-button>
          <var-button text type="primary" @click="timeConfirm"> {{ confirmButtonText }}</var-button>
        </div>
      </var-tab-item>
    </var-tabs-items>
  </div>
</template>

<script lang="ts">
import VarTimePicker from '../time-picker'
import VarDatePicker from '../date-picker'
import { defineComponent, ref } from 'vue'
import { props } from './props'
import { createNamespace, call, formatElevation } from '../utils/components'
import Snackbar from '../snackbar'
import dayjs from 'dayjs/esm'

export default defineComponent({
  name: 'VarDateTimePicker',
  components: {
    VarTimePicker,
    VarDatePicker,
  },
  props,
  setup(props) {
    const { name, n, classes } = createNamespace('date-time-picker')
    const activeTab = ref(+props.active)

    const cancel = () => {
      call(props.onCancel)
    }

    const currentDate = props.modelValue ? props.modelValue : dayjs().format('YYYY-MM-DD HH:MM')
    const [currentDay, currentTime] = currentDate.split(' ')
    const date1 = ref<string | undefined>(currentDay)
    const date2 = ref<string | undefined>(currentTime)

    const nextStep = () => {
      activeTab.value += 1
      console.log(activeTab.value)
    }
    const prevStep = () => {
      activeTab.value -= 1
    }

    const timeConfirm = () => {
      Snackbar(`${date1.value}  ${date2.value}`)
      call(props.onConfirm, `${date1.value} ${date2.value}`)
    }

    return {
      activeTab,
      cancel,
      name,
      n,
      classes,
      date1,
      date2,
      timeConfirm,
      nextStep,
      prevStep,
      formatElevation,
    }
  },
})
</script>

<style lang="less">
@import './dateTimePicker.less';
@import '../styles/elevation';
</style>
