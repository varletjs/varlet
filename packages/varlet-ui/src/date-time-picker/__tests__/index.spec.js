import DateTimePicker from '..'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'
import VarDateTimePicker from '../DateTimePicker.vue'
import { expect, vi } from 'vitest'
import { delay } from '../../utils/test'

test('test date-time-picker plugin', () => {
  const app = createApp({}).use(DateTimePicker)
  expect(app.component(DateTimePicker.name)).toBeTruthy()
})

describe('test dateTimePicker style and date', () => {
  test('test dateTimePicker style and date', async () => {
    const template = `
      <var-date-time-picker v-model="date"/>
    `
    const wrapper = mount({
      components: {
        [VarDateTimePicker.name]: VarDateTimePicker,
      },
      data() {
        return {
          date: '2021-04-08 12:22',
        }
      },
      template,
    })

    await delay(0)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
