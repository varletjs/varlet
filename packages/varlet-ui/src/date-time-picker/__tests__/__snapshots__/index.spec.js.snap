// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test dateTimePicker style and date > test dateTimePicker style and date 1`] = `
"<div class=\\"var-date-time-picker\\">
  <var-tabs-items active=\\"0\\">
    <var-tab-item>
      <div class=\\"var-date-picker\\">
        <div class=\\"var-date-picker__title\\">
          <div class=\\"var-date-picker__title-year\\">2021</div>
          <div class=\\"var-date-picker__title-date var-date-picker__title-date--active\\">
            <transition-stub name=\\"var-date-picker-translatey\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div>04-08 星期四</div>
            </transition-stub>
          </div>
        </div>
        <div class=\\"var-date-picker__body\\">
          <transition-stub name=\\"var-date-picker-panel-fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
            <div class=\\"var-day-picker\\">
              <div class=\\"var-day-picker__content\\">
                <div class=\\"var-picker-header\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--round\\" style=\\"filter: opacity(0.54);\\" type=\\"button\\">
                    <!--v-if-->
                    <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-chevron-left\\" style=\\"transition-duration: 0ms;\\"></i></div>
                    <div class=\\"var-hover-overlay\\"></div>
                  </button>
                  <div class=\\"var-picker-header__value\\">
                    <transition-stub name=\\"var-date-picker-translatex\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
                      <div>2021 四月</div>
                    </transition-stub>
                  </div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--round\\" style=\\"filter: opacity(0.54);\\" type=\\"button\\">
                    <!--v-if-->
                    <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-chevron-right\\" style=\\"transition-duration: 0ms;\\"></i></div>
                    <div class=\\"var-hover-overlay\\"></div>
                  </button>
                </div>
                <transition-stub name=\\"var-date-picker-translatex\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
                  <div>
                    <ul class=\\"var-day-picker__head\\">
                      <li>日</li>
                      <li>一</li>
                      <li>二</li>
                      <li>三</li>
                      <li>四</li>
                      <li>五</li>
                      <li>六</li>
                    </ul>
                    <ul class=\\"var-day-picker__body\\">
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--disabled var-button--text-primary var-button--text var-button--text-disabled var-button--round var-day-picker__button\\" type=\\"button\\" disabled=\\"\\" var-day-picker-cover=\\"\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\"></div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--disabled var-button--text-primary var-button--text var-button--text-disabled var-button--round var-day-picker__button\\" type=\\"button\\" disabled=\\"\\" var-day-picker-cover=\\"\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\"></div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--disabled var-button--text-primary var-button--text var-button--text-disabled var-button--round var-day-picker__button\\" type=\\"button\\" disabled=\\"\\" var-day-picker-cover=\\"\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\"></div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--disabled var-button--text-primary var-button--text var-button--text-disabled var-button--round var-day-picker__button\\" type=\\"button\\" disabled=\\"\\" var-day-picker-cover=\\"\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\"></div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">1</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">2</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">3</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">4</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">5</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">6</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">7</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2 var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"false\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">8</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">9</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">10</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">11</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">12</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">13</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">14</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">15</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">16</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">17</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">18</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">19</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">20</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">21</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">22</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">23</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">24</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">25</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">26</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">27</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">28</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">29</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                      <li><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-primary var-button--text var-button--round var-day-picker__button var-day-picker__button--usable\\" type=\\"button\\" var-day-picker-cover=\\"\\" var-date-picker-color-cover=\\"true\\">
                          <!--v-if-->
                          <div class=\\"var-button__content\\">30</div>
                          <div class=\\"var-hover-overlay\\"></div>
                        </button></li>
                    </ul>
                  </div>
                </transition-stub>
              </div>
            </div>
          </transition-stub>
        </div>
      </div>
      <div class=\\"var-date-time-picker__body\\">
        <var-button text=\\"\\">cancel</var-button>
        <var-button text=\\"\\" type=\\"primary\\">next</var-button>
      </div>
    </var-tab-item>
    <var-tab-item>
      <div class=\\"var-time-picker\\">
        <div class=\\"var-time-picker__title\\">
          <div class=\\"var-time-picker__title-time\\">
            <div class=\\"var-time-picker__title-btn var-time-picker__title-btn--active\\">12</div><span>:</span>
            <div class=\\"var-time-picker__title-btn\\">22</div>
            <!--v-if-->
            <!--v-if-->
          </div>
          <div class=\\"var-time-picker__title-ampm\\">
            <div class=\\"var-time-picker__title-btn var-time-picker__title-btn--active\\">AM</div>
            <div class=\\"var-time-picker__title-btn\\">PM</div>
          </div>
        </div>
        <div class=\\"var-time-picker__body\\">
          <div class=\\"var-time-picker__clock-container\\">
            <transition-stub name=\\"var-time-picker-panel-fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div class=\\"var-time-picker__clock\\">
                <div class=\\"var-time-picker__clock-hand\\" style=\\"transform: rotate(0deg); height: calc(50% - 4px);\\"></div>
                <div class=\\"var-time-picker__clock-item var-time-picker__clock-item--active\\" style=\\"left: 50%; top: 0%;\\">12</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 75%; top: 6.698729810778064%;\\">1</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 93.30127018922192%; top: 25%;\\">2</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 100%; top: 50%;\\">3</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 93.30127018922194%; top: 74.99999999999999%;\\">4</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 75.00000000000003%; top: 93.30127018922192%;\\">5</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 50%; top: 100%;\\">6</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 25.00000000000001%; top: 93.30127018922194%;\\">7</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 6.698729810778076%; top: 75.00000000000003%;\\">8</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 0%; top: 50.000000000000014%;\\">9</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 6.6987298107780475%; top: 25.000000000000032%;\\">10</div>
                <div class=\\"var-time-picker__clock-item\\" style=\\"left: 24.99999999999998%; top: 6.698729810778076%;\\">11</div>
                <!--v-if-->
              </div>
            </transition-stub>
          </div>
        </div>
      </div>
      <div class=\\"var-date-time-picker__body\\">
        <var-button text=\\"\\" type=\\"primary\\">prev</var-button>
        <var-button text=\\"\\">cancel</var-button>
        <var-button text=\\"\\" type=\\"primary\\">submit</var-button>
      </div>
    </var-tab-item>
  </var-tabs-items>
</div>"
`;
